### PR TITLE
Gradebookidentityaddition

### PR DESCRIPTION
**Pull request by dtracers**
Adds Identity code to the gradebook

depends on  #482

### CODE REVIEW

**Comment by dtracers** [comment]
I think this branch needs more unit tests for getting a grade of the different types

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java [hunk 1]
```diff
@@ -551,7 +554,7 @@ void insertSubmission(final String userId, String authId, String problemId, Stri
      * @throws DatabaseAccessException
      *         Thrown if grades are not found in the database.
      */
-    List<ProtoGrade> getAllAssignmentGradesStudent(final String courseId, final String authId)
+    List<ProtoGrade> getAllAssignmentGradesStudent(final String courseId, final String authId, String userId)
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java]: make userId a final

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java [hunk 2]
```diff
@@ -593,7 +599,7 @@ void insertSubmission(final String userId, String authId, String problemId, Stri
      * @throws DatabaseAccessException
      *         Thrown if a grade is not found in the database matching the requested parameters.
      */
-    ProtoGrade getGrade(final String authId, final ProtoGrade gradeData)
+    ProtoGrade getGrade(final String userId, final String authId, final ProtoGrade gradeData)
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java]: please put userId after authId like in the previous method

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java [hunk 3]
```diff
@@ -607,6 +613,19 @@ ProtoGrade getGrade(final String authId, final ProtoGrade gradeData)
...
+     * Gets the username for the userId.
+     *
+     * @param userId The userId that is being exchanged for the userId.
+     * @param authId The authentication of the user that is exchanging the userId.
+     * @param courseId The course the username is being asked for.
+     * @return The username if the permissions are successful.
+     * @throws DatabaseAccessException Thrown if the user does not exist or there are problems getting the username.
+     * @throws AuthenticationException Thrown if the user does not have the authentication to get the username.
+     */
+    String getUserNameForIdentity(final String userId, final String authId, final String courseId)
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java]: In getAllAssignmentGradesStudent you put userId last
pick an order and stick to it so it is constant for all methods

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/GradeManager.java [hunk 4]
```diff
@@ -99,6 +99,8 @@ private GradeManager() {
      *         The id used to authenticate the user adding the grade to ensure the user has valid permission.
      * @param grade
      *         The ProtoObject representing the grade to be added.
+     *         Assumptions:
+     *              The userId is the hashed userId.
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/GradeManager.java]: wait waht?? what was the original userId then?

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java [hunk 5]
```diff
@@ -597,18 +598,29 @@ public void addGrade(final String authId, final ProtoGrade grade) throws Authent
     }
 
     @Override
-    public ProtoGrade getGrade(final String authId, final ProtoGrade gradeData) throws AuthenticationException, DatabaseAccessException {
-        return GradeManager.getGrade(auth, database, authId, authId, gradeData);
+    public ProtoGrade getGrade(final String userId, final String authId, final ProtoGrade gradeData)
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java]: put user id after auithid plzzzz

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java (hunk 2 above)
Comment (dtracers) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java]: actually I believe the userId always goes before the authId

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java [hunk 6]
```diff
@@ -597,18 +598,29 @@ public void addGrade(final String authId, final ProtoGrade grade) throws Authent
...
+                    .setKey(stringStringEntry.getKey())
+                    .setValue(stringStringEntry.getValue())
+                    .build());
+        }
+        return builder.build();
     }
 
     @Override
-    public List<String> getCourseRoster(final String authId, final String courseId)
+    public String getUserNameForIdentity(final String userId, final String authId, final String courseId)
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java]: put userId at the end or change the others to userId in the beginning

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingRequestHandler.java [hunk 7]
```diff
@@ -56,14 +56,15 @@ private GradingRequestHandler() {
      *
      * @param institution The database interface.
      * @param request The request being sent.
-     * @param authId The Id of the user who sent the request.
+     * @param authId The Id of the user who sent the request used for authentication purposes.
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingRequestHandler.java]: make sure you update the doc of @param authId for all other methods that contain authId

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingRequestHandler.java [hunk 8]
```diff
@@ -76,7 +77,7 @@ private GradingRequestHandler() {
         if (instructor && allGrades) {
             returnList = institution.getAllAssignmentGradesInstructor(request.getItemId(COURSE_INDEX), authId);
         } else if (student && allGrades) {
-            returnList = institution.getAllAssignmentGradesStudent(request.getItemId(COURSE_INDEX), authId);
+            returnList = institution.getAllAssignmentGradesStudent(request.getItemId(COURSE_INDEX), authId, userId);
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingRequestHandler.java]: change the method that you use here to take userId before authId

**Pull request merged**